src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala

```diff
@@ -0,0 +1,130 @@
+      val profilePropertyMap = profile.propertyValueMap
+      val propertyUpdates = updateUserProperties(userInfo, profilePropertyMap)
+      val profileResponse: Future[PerRequestMessage] = propertyUpdates flatMap { responses =>
+        val allSucceeded = responses.forall { _.status.isSuccess }
+        allSucceeded match {
+          case true =>
+            val kv2 = FireCloudKeyValue(Some("isRegistrationComplete"), Some("true"))
+            val completionUpdate = pipeline {
+              Post(UserService.remoteSetKeyURL, ThurloeKeyValue(Some(userInfo.getUniqueId), Some(kv2)))
+            }
+            completionUpdate.flatMap { response =>
+              response.status match {
+                case x if x.isSuccess => checkUserInRawls(requestContext)
+                case _ => Future(RequestCompleteWithErrorReport(response.status, response.toString))
+              }
+            } recover { case e: Throwable => RequestCompleteWithErrorReport(InternalServerError, e.getMessage) }
+          case false =>
+            val errors = responses.filterNot(_.status == OK) map { e => (e, ErrorReport.tryUnmarshal(e) ) }
+            val errorReports = errors collect { case (_, Success(report)) => report }
+            val missingReports = errors collect { case (originalError, Failure(_)) => originalError }
+            val errorMessage = {
+              val baseMessage = "%d failures out of %d attempts saving profile.  Errors: %s"
+                .format(profilePropertyMap.size, errors.size, errors mkString ",")
+              if (missingReports.isEmpty) baseMessage
+              else {
+                val supplementalErrorMessage = "Additionally, %d of these failures did not provide error reports: %s"
+                  .format(missingReports.size, missingReports mkString ",")
+                baseMessage + "\n" + supplementalErrorMessage
+              }
+            }
+            Future(RequestCompleteWithErrorReport(InternalServerError, errorMessage, errorReports))
+        }
+      } recover { case e: Throwable => RequestCompleteWithErrorReport(InternalServerError, e.getMessage) }
+
+      profileResponse pipeTo context.parent
+
+  }
+
+  def updateUserProperties(userInfo: UserInfo, profilePropertyMap: Map[String, String]): Future[List[HttpResponse]] = {
+    val pipeline = authHeaders(requestContext) ~> sendReceive
+    val propertyPosts = profilePropertyMap map {
+      case (key, value) =>
+        pipeline {
+          Post(UserService.remoteSetKeyURL,
+            ThurloeKeyValue(
+              Some(userInfo.getUniqueId),
+              Some(FireCloudKeyValue(Some(key), Some(value)))
+          ))
+        }
+    }
+    Future sequence propertyPosts.toList
+  }
+
+  def checkUserInRawls(requestContext: RequestContext): Future[PerRequestMessage] = {
+    val pipeline = authHeaders(requestContext) ~> sendReceive
+    pipeline { Get(UserService.rawlsGetUserURL) } flatMap { response =>
```

Comment:
There is a reason. Map-ping over a future will give you another future back which works well if you have a single 'stack' of futures to deal with. If there are conditional or nested futures that need to be made, then you get this ugly `Future[Future[PerRequest]]` thing when working in a map. To avoid that, you have to flatMap instead. I think we need to have an in-person code-discussion for this scenario.